### DISCUSSION TITLE
Rethinking the Freeze Behavior on Fabric and Paper

### DISCUSSION THREAD

**Discussion by hirbod** [react-navigation/react-navigation#12337] [Ideas]
Hey @satya164  and @WoLewicki ,

Building on our earlier discussion in the PR comments about freezing screens, I wanted to start a focused thread about this since it's tied to broader behavior we've observed.

Currently, the logic to freeze inactive screens doesn't apply to Fabric in specific cases, as noted here:

```javascript
const shouldFreeze = isFabric()
  ? !isPreloaded && !isFocused && !isBelowFocused;
```

This approach made me think about a related issue I've noticed with NativeWind's active classes even on Paper. When navigating, some animated states (e.g., scaling an image on press or even setting a background color) persist unexpectedly because of freeze. It can feel reminiscent of Cupertino-style lists where active items remain active until you return. While this is visually interesting, it creates inconsistencies that might not always be intuitive. 

I've attached two videos demonstrating the behavior:

https://github.com/user-attachments/assets/aab16d10-6e26-4bb0-a03d-2dacaea6c32a

https://github.com/user-attachments/assets/82f29c42-3642-4c93-b48a-da3f80b9adcf

Should we explore a unified strategy for both Fabric and Paper to handle freezing and animated states more predictably? 
I am coming more and more to the conclusion that the direct back screen should not be frozen, or only after `transitionEnd` and should unfreeze the moment we try to pop or use a gesture (`onTransitionStart`).

React Freeze is definitely reacting a bit too soon, as you can see in my videos.

Looking forward to your thoughts!

**Comment by WoLewicki** [comment]
I think it is good point to maybe enable something like freezing on demand.

**Comment by WoLewicki** [comment]
> direct back screen should not be frozen

But then you can pop 2 screens at once and will see the same effect so I am not sure if it is doable in an easy way, I might be wrong here though.

**Reply by hirbod** [reply]
I think it's not true since the animation didn't freeze in the first place :D.  
Either way, I believe the correct behavior is to freeze on `onTransitionEnd` and unfreeze on `onTransitionStart` (or abort), not while in-flight.  

Edit:
With that approach, the back screen can remain frozen. The downside could be that elements start shifting, moving, or re-rendering the moment a user drags the screen to peek at the back screen.

Edit:
So coming full circle. The back screen should not be frozen :D

**Comment by satya164** [comment]
I think first thing would be to use `react-freeze` directly in React Navigation instead of through screens so we have better flexibility.

Freezing is similar to the existing behavior of detaching inactive screens, so some of those APIs could be mirrored and the behavior should be matched.

`freezePreviousScreen` similar to `detachPreviousScreen` for screens below transparent modals, or screens which may need to stay animating - and we default them to not frozen for transparent modals.

In the default behavior, freeze for screen below the active screen should only activate if `freezeOnBlur` is set to `true` and no transition is happening.

> So coming full circle. The back screen should not be frozen

It depends on the use case. Not everyone has a transparent modal, or animation in the previous screen they want running and may prefer to freeze instead. Not freezing during transition should already handle many cases.

I think in addition to adjusting the default behavior, the API should provide more control over when freeze happens so it can address different use cases.

**Comment by satya164** [comment]
Update on this: We're dropping the `ScreenContainer` usage entirely in React Navigation 8, and consequently dropping `react-freeze`. We [will use `React.Activity` instead](https://reactnavigation.org/blog/2026/03/10/react-navigation-8.0-march-progress#new-inactivebehavior-option-for-all-navigators).

This integration lives in React Navigation, so we have complete control over how it works, so we can avoid pausing screens if they are preloaded etc. We also delay pausing screens so it doesn't interrupt animations. It can also be disabled on a per screen basis.

Note that `React.Activity` is similar but not the same as `react-freeze`. While freeze suspends components so they can't re-render, pausing screen with `Activity` means it unmounts effects, so effectively your subscriptions get cleaned up which means screen won't normally re-render, but it's not frozen and re-render is still possible (e.g. due to context update).
